pkg/metrics/metrics.go

```diff
@@ -0,0 +1,312 @@
+		Handler:        mux,
+		ReadTimeout:    10 * time.Second,
+		WriteTimeout:   10 * time.Second,
+		MaxHeaderBytes: 1 << 20,
+	}
+
+	go func() {
+		klog.Infof("metric server listening at %q", mm.metricsEndpoint)
+		if err := metricServer.ListenAndServe(); err != nil {
+			klog.Errorf("failed to start metric server at specified endpoint %q and path %q: %v", mm.metricsEndpoint, metricsPath, err)
+		}
+	}()
+}
+
+// RegisterMetricsCollector registers the metrics collector. It is idempotent to register the same collector.
+func (mm *manager) RegisterMetricsCollector(targetPath, podNamespace, podName, bucketName string) {
+	emptyDirBasePath, err := util.PrepareEmptyDir(targetPath, false)
+	if err != nil {
+		klog.Errorf("failed to register metrics collector for pod %v/%v, bucket %q: %v", podNamespace, podName, bucketName, err)
+
+		return
+	}
+
+	podUID, volumeName, _ := util.ParsePodIDVolumeFromTargetpath(targetPath)
+	c := NewTextFileCollector(emptyDirBasePath, podUID, volumeName, map[string]string{
+		"pod_name":       podName,
+		"namespace_name": podNamespace,
+		"volume_name":    volumeName,
+		"bucket_name":    bucketName,
+		"pod_uid":        podUID,
+	})
+	if err := mm.registry.Register(c); err != nil && !strings.Contains(err.Error(), prometheus.AlreadyRegisteredError{}.Error()) {
+		klog.Errorf("failed to register metrics collector for pod  %v/%v, volume %q, bucket %q: %v", podNamespace, podName, volumeName, bucketName, err)
+	}
+}
+
+// UnregisterMetricsCollector unregisters the metrics collector.
+func (mm *manager) UnregisterMetricsCollector(targetPath string) {
+	podUID, volumeName, _ := util.ParsePodIDVolumeFromTargetpath(targetPath)
+
+	// textFileCollector uses a hash of pod UID and volume name as an identifier.
+	c := NewTextFileCollector("", podUID, volumeName, nil)
+	if ok := mm.registry.Unregister(c); !ok {
+		klog.Infof("Unregister metrics collector for targetPath %q is not needed since the collector is not registered", targetPath)
+	}
+}
+
+type textFileCollector struct {
+	path        string
+	constLabels map[string]string
+	podUID      string
+	volumeName  string
+}
+
+// NewTextFileCollector returns a new Collector exposing metrics read from the give path.
+func NewTextFileCollector(path, podUID, volumeName string, labels map[string]string) prometheus.Collector {
+	c := &textFileCollector{
+		path:        path,
+		constLabels: labels,
+		podUID:      podUID,
+		volumeName:  volumeName,
+	}
+
+	return c
+}
+
+// Describe emits the description of metrics.
+// Prometheus Registry relies on this func to identify collectors.
+func (c *textFileCollector) Describe(ch chan<- *prometheus.Desc) {
+	// Collector id is a hash of the values of the ConstLabels and fqName.
+	ch <- prometheus.NewDesc("gke_gcsfuse_csi_metric", "GKE GCSFuse CSI metric.", nil, map[string]string{"pod_uid": c.podUID, "volume_name": c.volumeName})
+}
+
+// Collect emits metrics.
+func (c *textFileCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
thinking a version skew case. Say the csi driver pod upgraded to the version which supports metrics, and the user pod is still using the old sidecar which does not export metrics, this code would fail in ProcessMetricsFile() as it will fail to find gen file. The code would keep reporting error logs. Which may be fine, but something to call out in playbooks, on how to filter out if its a genuine issue or a false positive

Reply:
Will downgrade the error to warning and open an item to document this false positive scenario.

Reply:
Done.